src/nc_monitor.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * twemproxy - A fast and lightweight proxy for memcached protocol.
+ * 
+ * Copyright (C) 2021, wei huang <wei.kukey@gmail.com>
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <nc_monitor.h>
+#include <nc_rbtree.h>
+
+struct rbtree monitor_tree;
+struct rbnode monitor_sentinel_node;
+
+void monitor_init()
+{
+    rbtree_init(&monitor_tree, &monitor_sentinel_node);
+}
+
+void monitor_deinit(struct context *ctx)
+{
+    struct rbnode *node = NULL;
+
+    while ((node = rbtree_min(&monitor_tree) != NULL))
+    {
+        struct conn *c = node->data;
+        if (event_del_conn(ctx->evb, c) != NC_OK)  {
+            log_warn("event del conn c %d failed, ignored: %s",
+                 c->sd, strerror(errno));
+        }
+        c->close(ctx, c);
+
+        rbtree_delete(&monitor_tree, node);
+        nc_free(node);
+    }
+}
+
+int mointor_is_empty()
```

Comment:
`monitor_is_empty` specifically should be `static inline` and go in nc_monitor.h due to being called on every processed message, others should stay in nc_monitor.c